eclair-core/src/main/scala/fr/acinq/eclair/router/RouteCalculation.scala

```diff
@@ -324,15 +324,24 @@ object RouteCalculation {
                                          routeParams: RouteParams,
                                          currentBlockHeight: Long): Either[RouterException, Seq[Route]] = {
     // We use Yen's k-shortest paths to find many paths for chunks of the total amount.
-    val numRoutes = {
-      val directChannelsCount = g.getEdgesBetween(localNodeId, targetNodeId).length
-      routeParams.mpp.maxParts.max(directChannelsCount) // if we have direct channels to the target, we can use them all
+    // When the recipient is a direct peer, we have complete visibility on our local channels so we can use more accurate MPP parameters.
+    val routeParams1 = {
+      case class DirectChannel(balance: MilliSatoshi, isEmpty: Boolean)
+      val directChannels = g.getEdgesBetween(localNodeId, targetNodeId).map(e => {
+        val balance = e.balance_opt.getOrElse(e.capacity.toMilliSatoshi)
```

Comment:
Why `getOrElse`? Shouldn't we always know the balance for local channels.

If so, we may as well set `0.msat` so that unexpected behavior is more evident?

Reply:
We always know the balance for local channels but only once they've come up online and sent their `LocalChannelUpdate`, there's a potential race here...but you're right that if we don't have the balance available yet, the path-finding algorithm will very likely fail because it will try to use too much liquidity from the channel, so your proposal of disabling those channels by setting balance to 0 msat makes sense.

Reply:
Actually that was incorrect: for local channels, we should always have balance information available (otherwise the channel will not yet be in the graph).

Since it's an `Option` maybe what we should do instead is `collect` only those for which it's defined?

Reply:
Done in https://github.com/ACINQ/eclair/pull/1599/commits/337ba17b5a57b987738f30f866338839843892ed